google-cloud-firestore/lib/google/cloud/firestore/transaction.rb

```diff
@@ -0,0 +1,640 @@
+        #   end
+        #
+        def update doc, data, update_time: nil
+          ensure_not_closed!
+
+          doc_path = coalesce_doc_path_argument doc
+
+          Convert.writes_for_update(doc_path, data,
+                                    update_time: update_time).each do |write|
+            @writes << write
+          end
+
+          nil
+        end
+
+        ##
+        # Deletes a document from the database.
+        #
+        # @param [String, DocumentReference] doc A string representing the
+        #   path of the document, or a document reference object.
+        # @param [Boolean] exists Whether the document must exist. When `true`,
+        #   the document must exist or an error is raised. Default is `false`.
+        #   Optional.
+        # @param [Time] update_time When set, the document must have been last
+        #   updated at that time. Optional.
+        #
+        # @example Delete a document using a document path:
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   firestore.transaction do |tx|
+        #     # Delete a document
+        #     tx.delete "cities/NYC"
+        #   end
+        #
+        # @example Delete a document using a document reference:
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a document reference
+        #   nyc_ref = firestore.doc "cities/NYC"
+        #
+        #   firestore.transaction do |tx|
+        #     # Delete a document
+        #     tx.delete nyc_ref
+        #   end
+        #
+        # @example Delete a document using `exists`:
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   firestore.transaction do |tx|
+        #     # Delete a document
+        #     tx.delete "cities/NYC", exists: true
+        #   end
+        #
+        # @example Delete a document using the `update_time` precondition:
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   last_updated_at = Time.now - 42 # 42 seconds ago
+        #
+        #   firestore.transaction do |tx|
+        #     # Delete a document
+        #     tx.delete "cities/NYC", update_time: last_updated_at
+        #   end
+        #
+        def delete doc, exists: nil, update_time: nil
+          ensure_not_closed!
+
+          doc_path = coalesce_doc_path_argument doc
+
+          @writes << Convert.write_for_delete(
+            doc_path, exists: exists, update_time: update_time)
+
+          nil
+        end
+
+        # @!endgroup
+
+        ##
+        # @private commit the transaction
+        def commit
+          ensure_not_closed!
+          return rollback if @writes.empty?
+          @closed = true
+          return if @writes.empty?
```

Comment:
You need to commit even when you have no writes in order to release the read locks.

Reply:
Ick, I thought I had changed this. Fixed now.